File: vmsserver-go/w2cplatform/memvariables.go
package w2cplatform

// The config store in memory, with the same contract as the file one — and therefore right exactly when
// the contract suite is green against it:
//
//	CONTRACT_URL=memory:// go test ./w2cplatform -run Contract
//
// Lesson 3 said the store is a URL and nothing in the platform may know which backend answers it. That is
// a claim about design, and a design claim with one implementation is a hope. This is the second
// implementation: the same seven clauses, in memory, in a hundred lines.
//
// It is not a mock and it is not test scaffolding. It is registered on the same seam file:// is, chosen
// the same way (CONFIG_URL=memory://), and held to the same standard. What it is FOR: a dev box where
// nothing should survive a restart; a test that wants the real store and not a fake of it; and М11, where
// FakeVariables is this same thing with raft's answers — which is the point. A fake that implements a
// contract is a backend.
//
// A URL names a store, so two opens of one URL must be one store, as two opens of file://<dir> are:
//
//	memory://<name>   one store per name, per process. Three identities in one process — a controller, a
//	                  console and a worker in one dev binary — open the same name and see each other's
//	                  writes: sharing an address space is not sharing a writer, and the ACL still holds.
//	memory://         no name, so nothing to share by: a fresh private store on every open. That is what
//	                  the contract suite needs, since every clause starts from an empty store.
//
// Named stores live for the life of the process and are never collected. For a dev box and a test that is
// the whole of the requirement; anything that needs them collected wanted file://.

import (
	"fmt"
	"sort"
	"strconv"
	"strings"
	"sync"
)

var (
	namedMu  sync.Mutex
	namedMem = map[string]*memState{}
)

func init() {
	// The seam takes a URL and an identity, never a class name.
	RegisterScheme("memory", func(url, writer string, acl map[string][]string) (Variables, error) {
		v := NewMemVariables()
		// `memory://<name>?max_bytes=<n>`: the ceiling is part of WHICH STORE THIS IS, so it belongs in
		// the URL beside the name, exactly as the backend itself does (Lesson 20).
		rest := strings.TrimPrefix(url, "memory://")
		name, query, _ := strings.Cut(rest, "?")
		for _, part := range strings.Split(query, "&") {
			if k, val, ok := strings.Cut(part, "="); ok && k == "max_bytes" {
				if n, err := strconv.Atoi(val); err == nil {
					v.Max = n
				}
			}
		}
		if name != "" {
			namedMu.Lock()
			if st, ok := namedMem[name]; ok {
				v.s = st
			} else {
				namedMem[name] = v.s
			}
			namedMu.Unlock()
		}
		if writer == "" {
			return v, nil
		}
		return v.AsWriter(writer, acl[writer]...), nil
	})
}

// MemVariables is one identity over a shared state, exactly as FileVariables is one identity over a
// directory.
type MemVariables struct {
	s      *memState
	Writer string // "who am I" for the ACL check; "" bypasses it
	Max    int    // a map in memory has no ceiling — but `memory://x?max_bytes=512` gives the contract
	//               suite a store that DOES, which is how clause 8 is exercised against something
	//               other than prose
}

func (m *MemVariables) MaxBytes() int { return m.Max }

type memEntry struct {
	items Items
	index int64
}

type memState struct {
	mu    sync.Mutex
	index int64
	items map[string]memEntry
	acl   map[string][]string
}

// NewMemVariables: an empty store. The index starts at 1000 and only goes up, as the file backend's
// counter does — so a CAS from before a restart still conflicts instead of quietly matching.
func NewMemVariables() *MemVariables {
	return &MemVariables{s: &memState{index: 1000, items: map[string]memEntry{}, acl: map[string][]string{}}}
}

// AsWriter is the same store seen through another identity, allowed only these prefixes.
func (m *MemVariables) AsWriter(writer string, allowed ...string) *MemVariables {
	m.s.mu.Lock()
	if allowed != nil {
		m.s.acl[writer] = allowed
	}
	m.s.mu.Unlock()
	return &MemVariables{s: m.s, Writer: writer, Max: m.Max}
}

func (m *MemVariables) refuse(path string) error {
	if m.Writer == "" || len(m.s.acl) == 0 {
		return nil
	}
	if !Allowed(path, m.s.acl[m.Writer]) {
		return fmt.Errorf("%w: %s may not write %s", ErrForbidden, m.Writer, path)
	}
	return nil
}

// indexOf, held under the lock: Absent for a path that is not there.
func (m *memState) indexOf(path string) Index {
	e, ok := m.items[path]
	if !ok {
		return Absent
	}
	return Idx(strconv.FormatInt(e.index, 10))
}

// Get returns a COPY of the row — contract clause 7. A caller that mutated what it read would otherwise
// have edited the store with no index: no CAS, no conflict, no version, and the next reader sees a change
// nobody can point at. The file backend gets this for free by parsing JSON every time; here it is a
// decision, and the suite is what keeps it made.
func (m *MemVariables) Get(path string) (Items, Index, error) {
	if _, err := safe(path); err != nil {
		return nil, Absent, err
	}
	m.s.mu.Lock()
	defer m.s.mu.Unlock()
	e, ok := m.s.items[path]
	if !ok {
		return nil, Absent, nil
	}
	out := make(Items, len(e.items))
	for k, v := range e.items {
		out[k] = v
	}
	return out, Idx(strconv.FormatInt(e.index, 10)), nil
}

func (m *MemVariables) Put(path string, items Items, cas Index) (Index, error) {
	if _, err := safe(path); err != nil {
		return Absent, err
	}
	if err := Check(path, ItemsBytes(items), m.Max); err != nil {
		return Absent, err
	}
	m.s.mu.Lock()
	defer m.s.mu.Unlock()
	if err := m.refuse(path); err != nil {
		return Absent, err
	}
	current := m.s.indexOf(path)
	if cas.Conditional() && cas != current {
		return Absent, fmt.Errorf("%w: %s: cas=%s but the store is at %s", ErrConflict, path, cas, current)
	}
	m.s.index++
	cp := make(Items, len(items)) // the other half of clause 7: put is given a row and must not go on sharing it
	for k, v := range items {
		cp[k] = v
	}
	m.s.items[path] = memEntry{cp, m.s.index}
	return Idx(strconv.FormatInt(m.s.index, 10)), nil
}

func (m *MemVariables) List(prefix string) ([]string, error) {
	m.s.mu.Lock()
	defer m.s.mu.Unlock()
	out := []string{}
	for p := range m.s.items {
		if strings.HasPrefix(p, prefix) {
			out = append(out, p)
		}
	}
	sort.Strings(out)
	return out, nil
}

// Delete: the same CAS check as Put and — as on the file backend — no ACL check. A delete burns an index
// too, for the same reason a write does.
func (m *MemVariables) Delete(path string, cas Index) error {
	if _, err := safe(path); err != nil {
		return err
	}
	m.s.mu.Lock()
	defer m.s.mu.Unlock()
	current := m.s.indexOf(path)
	if cas.Conditional() && cas != current {
		return fmt.Errorf("%w: %s: cas=%s but the store is at %s", ErrConflict, path, cas, current)
	}
	delete(m.s.items, path)
	m.s.index++
	return nil
}

File: vmsserver/w2cplatform/memvariables.py
"""The config store in memory, with the same contract as the file one — and
therefore right exactly when the contract suite is green against it."""
# ================================================================================================
# NOTES — what every part of this file does and why (kept beside the code, not in a separate document)
# ================================================================================================
# # memvariables.py — `memory://`, the second backend, and what having two of them proves
#
# **Role in the module.** Lesson 3 said the store is a URL and nothing in the platform may know which
# backend answers it. That is a claim about design, and a design claim with one implementation is a hope.
# This is the second implementation: the same five clauses, in memory, in fifty lines.
#
# It is not a mock and it is not test scaffolding. It is registered on the same seam `file://` is, chosen
# the same way (`CONFIG_URL=memory://`), and held to the same standard — `test_variables_contract.py` runs
# against it unchanged:
#
#     CONTRACT_URL=memory:// python3 tests/run.py
#
# What it is FOR, in order of how often it matters: a dev box where nothing should survive a restart; a
# test that wants the real store and not a fake of it; and М11, where `FakeVariables` is this same thing
# with raft's answers — which is the point. A fake that implements a contract is a backend.
#
# ## Naming, and why a bare `memory://` is different
# A URL names a store, so two opens of one URL must be one store — as two opens of `file://<dir>` are.
# - `memory://<name>` — one store per name, per process. Three identities in one process (a controller, a
#   console and a worker in one dev binary) open the same name and see each other's writes: sharing an
#   address space is not sharing a writer, and the ACL still holds between them.
# - `memory://` — no name, so nothing to share by: a fresh private store on every open. That is what the
#   contract suite needs, since every clause starts from an empty store.
#
# The named stores live for the life of the process and are never collected. For a dev box and a test that
# is the whole of the requirement; anything that needs them collected wanted `file://`.
#
# ## What it deliberately copies from FileVariables
# Not the implementation — the BEHAVIOUR, down to the details a caller can see: an index is never reused,
# not even by a path that was deleted and came back; `put` checks the ACL and
# `delete` does not; `get` of an absent path is `(None, 0)`; a bad path is refused by `safe_path` before
# anything else happens. Where those differ, the contract suite is what says so — which is the entire
# argument for having written the suite before the second backend.
# ================================================================================================
from __future__ import annotations

import threading

from .variables import Conflict, Forbidden, register_scheme, safe_path

# One store per name, per process. Module-level because that is what "per process" means; the lock is
# around the registry, not around a store — each store has its own.
_NAMED: dict[str, "_MemState"] = {}
_NAMED_LOCK = threading.Lock()


class _MemState:
    """What the handles share. A handle is an identity over this."""

    def __init__(self) -> None:
        self.lock = threading.Lock()
        self.index = 1000           # matched to the file store so two backends read alike; a store that dies
                                    # with its process has no "before the restart", so the value is cosmetic here
        self.items: dict[str, tuple[dict, int]] = {}
        self.acl: dict[str, list[str]] = {}


class MemVariables:
    # `writer` is this handle's identity (None means unrestricted); `acl` is `{writer: [prefixes]}`. Both
    # mean exactly what they mean in `FileVariables`.
    def __init__(self, state: "_MemState | None" = None, writer: str | None = None,
                 acl: dict[str, list[str]] | None = None):
        self._s = state or _MemState()
        self.writer = writer
        if acl:
            with self._s.lock:
                self._s.acl.update(acl)

    # The same store seen through another identity, allowed only these prefixes.
    def as_writer(self, writer: str, allowed: list[str]) -> "MemVariables":
        with self._s.lock:
            self._s.acl[writer] = allowed
        return MemVariables(self._s, writer)

    @property
    def acl(self) -> dict[str, list[str]]:
        return self._s.acl

    def _refuse(self, path: str) -> None:
        if self.writer is None or not self._s.acl:
            return
        allowed = self._s.acl.get(self.writer, [])
        if not any(path == p or (p.endswith("*") and path.startswith(p[:-1])) for p in allowed):
            raise Forbidden(f"{self.writer} may not write {path}")

    # `(items, index)`, or `(None, 0)` for a path never written. The items are a COPY: a caller that
    # mutated what it read would be editing the store without an index, which is the one thing CAS exists
    # to prevent — and on the file backend it is impossible, so it must be impossible here.
    def get(self, path: str) -> tuple[dict | None, int]:
        safe_path(path)
        with self._s.lock:
            e = self._s.items.get(path)
            return (dict(e[0]), e[1]) if e else (None, 0)

    def put(self, path: str, items: dict, cas: int | None = None) -> int:
        safe_path(path)
        self._refuse(path)
        with self._s.lock:
            current = self._s.items.get(path, (None, 0))[1]
            if cas is not None and cas != current:
                raise Conflict(f"{path}: cas={cas} but ModifyIndex={current}")
            self._s.index += 1
            self._s.items[path] = ({k: str(v) for k, v in items.items()}, self._s.index)
            return self._s.index

    # The same CAS check as `put`, and — as on the file backend — no ACL check.
    def delete(self, path: str, cas: int | None = None) -> None:
        safe_path(path)
        with self._s.lock:
            current = self._s.items.get(path, (None, 0))[1]
            if cas is not None and cas != current:
                raise Conflict(f"{path}: cas={cas} but ModifyIndex={current}")
            self._s.items.pop(path, None)
            self._s.index += 1          # a delete burns an index too: the file backend's does

    def list(self, prefix: str) -> list[str]:
        with self._s.lock:
            return sorted(p for p in self._s.items if p.startswith(prefix))


# The seam takes a URL and an identity, never a class name. `memory://` is all a process says to keep
# nothing across a restart.
def _open(url: str, writer: str | None = None, acl: dict[str, list[str]] | None = None) -> MemVariables:
    name = url[len("memory://"):] if url.startswith("memory://") else ""
    if not name:
        return MemVariables(None, writer, acl)          # no name, no sharing: private to this open
    with _NAMED_LOCK:
        state = _NAMED.setdefault(name, _MemState())
    return MemVariables(state, writer, acl)


register_scheme("memory", _open)
